File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-211617-6449/output/project/src/customerimporter/identification.py
from abc import abstractmethod

from customerimporter.imported_value import ImportedValue


class Identification(ImportedValue):

    # instance creation

    @classmethod
    def for_(cls, an_identification_type, a_number):
        return cls._type_that_can_handle(an_identification_type).with_value(a_number)

    # type

    @classmethod
    def can_handle(cls, an_identification_type):
        return cls.type_code() == an_identification_type

    @classmethod
    @abstractmethod
    def type_code(cls):
        pass

    # error messages

    @classmethod
    def invalid_type_error_description(cls):
        return "Invalid identification type"

    # number

    def cuit_number_if_none(self, a_none_block):
        return a_none_block()

    def dni_number_if_none(self, a_none_block):
        return a_none_block()

    def number(self):
        return self._value

    # testing

    def is_cuit(self):
        return False

    def is_dni(self):
        return False

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2022-1c-recuperatorio-parcial1/20260921-202259-53404/analysis/acceptance-project/src/main/java/customerimporter/Identification.java
package customerimporter;

import java.util.List;
import java.util.function.Function;
import java.util.function.Supplier;

/**
 * The identification of a {@link Customer}: knows its type code, its number and how to answer
 * the questions that only make sense for one kind of identification. Each subclass validates
 * its own numbers, so nobody has to ask which kind of identification it is dealing with.
 */
public abstract class Identification {

    private static final List<Kind> kinds = List.of(
        new Kind(DNI.TYPE_CODE, DNI::with),
        new Kind(CUIT.TYPE_CODE, CUIT::with));

    private final String number;

    // instance creation

    public static Identification of(String aTypeCode, String aNumber) {
        return kinds.stream()
            .filter(aKind -> aKind.handles(aTypeCode))
            .findFirst()
            .orElseThrow(() -> new RuntimeException(invalidTypeErrorDescription()))
            .create(aNumber);
    }

    // error messages

    public static String invalidTypeErrorDescription() {
        return "Invalid identification type";
    }

    // initialization

    protected Identification(String aNumber) {
        number = aNumber;
    }

    // accessing

    public String number() {
        return number;
    }

    public abstract String typeCode();

    // testing

    public boolean isIdentifiedBy(String aTypeCode, String aNumber) {
        return typeCode().equals(aTypeCode) && number.equals(aNumber);
    }

    public abstract boolean isCUIT();

    public abstract boolean isDNI();

    // evaluating

    public abstract Object cuitNumberIfNone(Supplier<Object> aNoneBlock);

    public abstract Object dniNumberIfNone(Supplier<Object> aNoneBlock);

    // kinds - private

    private record Kind(String typeCode, Function<String, Identification> creator) {

        boolean handles(String aTypeCode) {
            return typeCode.equals(aTypeCode);
        }

        Identification create(String aNumber) {
            return creator.apply(aNumber);
        }
    }
}
